File: src/main/java/me/taubsie/dungeonhub/application/command/commands/ScoreCommand.java
package me.taubsie.dungeonhub.application.command.commands;

import me.taubsie.dungeonhub.application.command.Command;
import me.taubsie.dungeonhub.application.command.CommandParameters;
import me.taubsie.dungeonhub.application.connection.dungeon_hub.ServerConnection;
import me.taubsie.dungeonhub.application.exceptions.InvalidOptionException;
import me.taubsie.dungeonhub.application.service.ApplicationService;
import me.taubsie.dungeonhub.common.model.score.ScoreModel;
import me.taubsie.dungeonhub.common.model.server.ServerModel;
import org.javacord.api.entity.message.MessageFlag;
import org.javacord.api.entity.user.User;
import org.javacord.api.event.interaction.SlashCommandCreateEvent;
import org.javacord.api.interaction.SlashCommandOption;
import org.javacord.api.interaction.SlashCommandOptionBuilder;
import org.javacord.api.interaction.SlashCommandOptionType;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

@CommandParameters(name = "score", description = "Use this to see the score of yourself or another user.")
public class ScoreCommand extends Command {
    @Override
    protected void executeCommand(SlashCommandCreateEvent slashCommandCreateEvent) {
        User userToCheck;
        try {
            userToCheck = getUserOption("user");
        }
        catch (InvalidOptionException invalidOptionException) {
            userToCheck = getUser();
        }

        List<ScoreModel> scores = ServerConnection.getInstance()
                .getScores(new ServerModel(getServer().getId()), userToCheck.getId())
                .orElse(new ArrayList<>());

        slashCommandCreateEvent
                .getSlashCommandInteraction()
                .createImmediateResponder()
                .setFlags(MessageFlag.EPHEMERAL)
                .addEmbed(ApplicationService.getInstance()
                        .getScoreCountMessage(userToCheck, getUser(), getServer(), scores))
                .respond();
    }

    @Override
    public List<SlashCommandOption> getSlashCommandOptions() {
        SlashCommandOption userOption = new SlashCommandOptionBuilder()
                .setType(SlashCommandOptionType.USER)
                .setName("user")
                .setDescription("The user to check the carries for.")
                .setRequired(false)
                .build();

        return Collections.singletonList(userOption);
    }
}

File: src/main/java/me/taubsie/dungeonhub/kord/application/commands/ScoreCommand.kt
package me.taubsie.dungeonhub.kord.application.commands

import com.kotlindiscord.kord.extensions.commands.Arguments
import com.kotlindiscord.kord.extensions.commands.converters.impl.optionalUser
import com.kotlindiscord.kord.extensions.extensions.Extension
import com.kotlindiscord.kord.extensions.extensions.ephemeralSlashCommand
import me.taubsie.dungeonhub.application.connection.dungeon_hub.DiscordServerConnection
import me.taubsie.dungeonhub.common.model.server.DiscordServerModel
import me.taubsie.dungeonhub.kord.application.loader.LoadExtension
import me.taubsie.dungeonhub.kord.application.service.ApplicationService

@LoadExtension
class ScoreCommand : Extension() {
    override val name = "score-command"

    override suspend fun setup() {
        ephemeralSlashCommand(::ScoreArguments) {
            name = "score"
            description = "Use this to see the score of yourself or another user."
            allowInDms = false

            action {
                respond {
                    val userToCheck = arguments.user ?: event.interaction.user

                    val scores = DiscordServerConnection.getInstance()
                        .getScores(DiscordServerModel(guild!!.id.value.toLong()), userToCheck.id.value.toLong())
                        .orElse(listOf())

                    embeds = mutableListOf(
                        ApplicationService.getScoreCountMessage(userToCheck, event.interaction.user, guild, scores)
                    )
                }
            }
        }
    }

    inner class ScoreArguments : Arguments() {
        val user by optionalUser {
            name = "user"
            description = "The user to check the carries for."
        }
    }
}